Tests for C++ custom autograd function API

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23628 Tests for C++ custom autograd function API**

Summary: More tests for autograd::Fuction based on python tests from test_autograd.py

Differential Revision: [D16600992](https://our.internmc.facebook.com/intern/diff/D16600992)

Commit:
Tests for C++ customm autograd function API

Summary: More tests for autograd::Fuction

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Update on "Tests for C++ custom autograd function API"


Summary: More tests for autograd::Fuction based on python tests from test_autograd.py

ezyang commented on test/cpp/api/autograd.cpp:
```diff
@@ -105,6 +106,50 @@ TEST(CustomAutogradTest, MarkNonDifferentiable) {
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+      Variable &grad_a = grad_output[0], &grad_b = grad_output[1];
```
Taking a mutable reference here is kind of goofy but I guess it does no harm in this particular block of code.

ezyang commented on test/cpp/api/autograd.cpp:
```diff
@@ -127,3 +172,254 @@ TEST(CustomAutogradTest, ReturnLeafInplace) {
+  ASSERT_VARIABLE_EQ(x.grad(), torch::ones(x.sizes()));
+}
+
+TEST(CustomAutogradTest, Reentrant) {
```
Wow! C++ reentrant backwards :3

Commit:
Update on "Tests for C++ custom autograd function API"


Summary: More tests for autograd::Fuction based on python tests from test_autograd.py

Differential Revision: [D16600992](https://our.internmc.facebook.com/intern/diff/D16600992)

yf225 commented on test/cpp/api/autograd.cpp:
```diff
@@ -148,3 +232,215 @@ TEST(CustomAutogradTest, SaveEmptyForBackward) {
+  };
+}
+
+TEST(CustomAutogeradTest, DepNoGrad) {
```
nit: "CustomAutogeradTest" -> "CustomAutogradTest"

Commit:
Update on "Tests for C++ custom autograd function API"


Summary: More tests for autograd::Fuction based on python tests from test_autograd.py

Differential Revision: [D16600992](https://our.internmc.facebook.com/intern/diff/D16600992)

Commit:
Update on "Tests for C++ custom autograd function API"


Summary: More tests for autograd::Fuction based on python tests from test_autograd.py

Differential Revision: [D16600992](https://our.internmc.facebook.com/intern/diff/D16600992)
